### ISSUE TITLE
Typed JSON and name clashes

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3143]
Relates to old #239, #240 and #299.

@chrhol: notice I wont even assign you, this is *not* something I expect you to solve. At least not in in the near future. :wink: 

Case: defining certain properties in typed JSON clash with properties in the underlying template. Just now I defined:

**ProductTypePartial.json**
```json
{
  "Id" : "",
  "Name" : "",
  "Description" : "",
  "Parent": ""
}
```

and got

```
json error : ViewModels\ProductTypePartial.json already contains a definition for 'Parent'
```

Ideally, we should try get away from it completely. But I think that will be hard. Second best: lets try document what property names are actually reserved like this, and try to produce an error that describe that, and keep a link to that list in the docs.

**Comment by per-samuelsson** [comment]
> Second best: lets try document what property names are actually reserved like this, and try to produce an error that describe that, and keep a link to that list in the docs.

Like:


> ScErrReservedJSONProperty: The name `Parent` is a reserved JSON property. Please name this property something else. A list of reserved properties can be found here: starcounter.io/json/reservednames.
```

**Comment by chrhol** [comment]
> Ideally, we should try get away from it completely. But I think that will be hard.

I think it will not just be hard, but rather impossible with our current solution, since this is a limitation in .Net (and I guess alot of other languages) and there will always be some clashes. The only option would be to totally rework how the generated code looks like (for example have a special namespace or special names for the generated properties). 

Worth to notice though is that the generated properties only calls underlying code for getting and setting values using delegates thar are created on each template for the TypedJson, which handles different kind of storage (bound, unbound, set directly on instance, ...) 

**ProductTypePartial.json.g.cs (snippet of generated code)**
```csharp
partial class ProductTypePartial {
    ...
    public string Name {
        get { return Template.Name.Getter(this); }
        set { Template.Name.Setter(this, value); }
    }
    ...
}
```

So maybe there is some smart way to handle name clashes?

> Second best: lets try document what property names are actually reserved like this, and try to produce an error that describe that, and keep a link to that list in the docs.

Good idea. The error is already thrown from our code today, when a template is created (AppTemplateFactory.cs, VerifyPropertyNames) but only contains a few reserved properties and not a good error-message (not even a proper errorcode)

**Comment by per-samuelsson** [comment]
> Good idea. The error is already thrown from our code today, when a template is created (AppTemplateFactory.cs, VerifyPropertyNames) but only contains a few reserved properties and not a good error-message (not even a proper errorcode)

Great. Simple as "small pancake" then! I volonteer to write docs about it; I'll assign this to me for that.

**Comment by Starcounter-Jack** [comment]
Maybe the reserved names should be less generic?

**Comment by per-samuelsson** [comment]
> Maybe the reserved names should be less generic?

Yeah, I know we did that for "Index" and "Name" earlier. Lets see what we can do about the ones we have still, and document them no matter what (even after we make them less generic).

The thing is, most typed JSON (I think) is maintain mainly as... well, JSON. Not as the typed instance. And at least not so when its about developers accessing built-in properties.

As an example, how many developers would access `Json.Parent`. That one, we could rename to `Json.ParentJson` for example, freeing `Parent` from being reserved.

**Comment by andwah** [comment]
Just ran into this problem myself. I had a property named `Parent` in my db class.

As i understand these names should be avoided as property names in your db class.
(From the current Json class)

- AutoRefreshBoundProperties
- BoolValue
- ChangeLog
- Data
- DecimalValue
- DoubleValue
- IndexInParent
- IntegerValue
- IsArray
- IsBool
- IsCodegenerated
- IsDecimal
- IsDirty
- IsDouble
- IsInteger
- IsObject
- IsPersistent
- IsPrimitive
- IsString
- Parent
- Root
- Session
- StringValue
- Template
- Transaction

**Comment by mmnosek** [comment]
It hit me also today. I've almost created a new issue :). I think that we should deal with this topic somehow. If we can't allow using this names we should at least:
* Change the error message from `The type 'PageClass' already contains a definition for 'SomeName'` to something more understandable. I've spent some time investigating why actually I have any other definition. It's very confusing.
* Document it clearly on ".io" site. I can create a separate issue for documenting it. Is the list provided by @andwah complete?

cc @per-samuelsson @chrhol

**Comment by andwah** [comment]
> Is the list provided by @andwah complete?

Not sure, i just check the `Json `class for possible clashes.
